File: src/main/java/com/tms/csp/transforms/Transformer.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;
import com.tms.csp.ast.PLConstants;

import javax.annotation.Nonnull;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.tms.csp.ssutil.Strings.getSimpleName;

public abstract class Transformer extends Transforms implements PLConstants {

    public static final Transformer FLATTEN_ANDS = new FlattenAnds();
    public static final Transformer FLATTEN_IMPS = new FlattenImps();
    public static final Transformer FLATTEN_RMPS = new FlattenRmps();
    public static final Transformer FLATTEN_ORS = new FlattenOrs();
    public static final Transformer OR_TO_AND = new OrToAnd();
    public static final Transformer AND_TO_BINARY = new AndsToBinary();
    public static final Transformer FLATTEN = new UntilStableTransformer(FLATTEN_ANDS, FLATTEN_ORS);


    public String name;

    public String getName() {
        if (name == null) {
            name = getSimpleName(this);
        }
        return name;
    }

    public Object howToHandle(@Nonnull Exp in) {
        return null;
    }


    @Nonnull
    public abstract Exp transform(@Nonnull Exp in);

    public Exp transformUntilStable(@Nonnull Exp in) {
        return transformUntilStable(this, in);
    }

    public List<Exp> transformArgs(List<Exp> args) {
        List<Exp> exps = transformArgs(this, args);
        if (exps == null) {
            throw new IllegalStateException();
        }
        return exps;
    }

    public Iterable<Exp> transformArgs2(Iterable<Exp> args) {
        Iterable<Exp> exps = transformArgs2(this, args);
        if (exps == null) {
            throw new IllegalStateException();
        }
        return exps;
    }


    static int outMatchesInCount = 0;

    @Nonnull
    public static Exp transformUntilStable(Transformer transformer, @Nonnull Exp in) {

        Exp out = transformer.transform(in);

        if (out == in) {
            return out;
        } else {
            return transformUntilStable(transformer, out);
        }
    }

    public static List<Exp> transformList(Transformer transformer, List<Exp> list) {
        return ListTransform.transform(list, transformer);
    }

    /**
     * Leave Exp.type the same
     */
    protected Exp transformArgsOnly(Exp in) {
        checkNotNull(in);
        if (in.isConstant() || in.isLit()) {
            return in;
        } else if (in.isNegComplex()) {
            Exp argBefore = in.getArg();
            Exp argAfter = this.transform(argBefore);
            if (argAfter == argBefore) {
                return in;
            } else {
                return argAfter.flip();
            }
        } else if (in.isPosComplex()) {
            Iterable<Exp> inArgs = in.args();
            Iterable<Exp> outArgs = transformArgs2(inArgs);

            Exp retVal;
            if (outArgs == inArgs) {
                retVal = in;
            } else {
                retVal = in.asPosComplex().newComplex(outArgs);
            }

            checkNotNull(retVal);

            return retVal;
        } else {
            throw new IllegalArgumentException();
        }


    }

    public static List<Exp> transformArgs(Transformer transformer, List<Exp> argsIn) {
        List<Exp> listOut = ListTransform.transform(argsIn, transformer);
        if (argsIn == listOut) {
            return argsIn;
        } else {
            return listOut;
        }
    }

    public static Iterable<Exp> transformArgs2(Transformer transformer, Iterable<Exp> argsIn) {
        Iterable<Exp> listOut = ListTransform.transform2(argsIn, transformer);
        if (argsIn == listOut) {
            return argsIn;
        } else {
            return listOut;
        }
    }

//    public ExpSet applyDelta(Exp[] delta) {
//        if (delta == null) {
//            return this;
//        }
//
//        Exp[] ret = new Exp[size()];
//        for (int i = 0; i < size(); i++) {
//            if (delta[i] != null) {
//                ret[i] = delta[i];
//            } else {
//                ret[i] = get(i);
//            }
//        }
//
//
//        return new NSet(ret);
//    }


    public static final Transformer IDENTITY = new IdentityTransformer();


    public static final Transformer PUSH_NOTS_IN = new UntilStableTransformer(new CompoundTransformer(
            new PushNotsIn(),
            FLATTEN,
            new OneArgOr(), new OneArgAnd()
    ));

    public static final Transformer PUSH_ANDS_OUT = new PushAndsOut();


    /**
     * BNF:mBase Normal Form
     * Allow:
     * And(N)
     * Not
     * Or(N)
     * Var
     */
    public static final Transformer BNF = new UntilStableTransformer("BNF", new CompoundTransformer(
            new FlattenImps(),
            new FlattenRmps(),
            new XorsToCnf(),
            new ConflictsToCnf(),
            new IffToCnf(),
            new ImpsToCnf(),
            new RmpsToCnf(),
            new FlattenOrs(),
            new FlattenAnds(),
            new OneArgOr(),
            new OneArgAnd()
    ));

    public static final Transformer BNF_KEEP_XORS = new UntilStableTransformer("BNF", new CompoundTransformer(
            new FlattenImps(),
            new FlattenRmps(),
            new ConflictsToCnf(),
            new IffToCnf(),
            new ImpsToCnf(),
            new RmpsToCnf(),
            new FlattenOrs(),
            new FlattenAnds(),
            new OneArgOr(),
            new OneArgAnd()
    ));


    public static final Transformer BNF_TO_NNF = new UntilStableTransformer("BNF_TO_NNF", PUSH_NOTS_IN, FLATTEN);

    /**
     * Only allow:
     * Same as BNF,but only _complexVars can be negated
     */
    public static final Transformer NNF = new UntilStableTransformer("NNF", BNF, BNF_TO_NNF);

    public static final Transformer BNF_TO_AIG = new UntilStableTransformer("AIG",
            OR_TO_AND,
            AND_TO_BINARY
    );


    public static final Transformer NNF_TO_CNF = new UntilStableTransformer("NNF_TO_CNF",
            new PushAndsOut(),
            new OneArgOr(),
            new OneArgAnd(),
            FLATTEN
    );

    public static final Transformer CNF = new UntilStableTransformer("CNF", NNF_TO_CNF, NNF);

    public static final Transformer AIG = new CompoundTransformer("AIG", BNF, BNF_TO_AIG);


}

File: src/main/java/com/smartsoft/csp/transforms/Transformer.kt
package com.smartsoft.csp.transforms

import com.smartsoft.csp.ast.Exp
import com.smartsoft.csp.ast.PLConstants

import com.google.common.base.Preconditions.checkNotNull
import com.smartsoft.csp.ssutil.Strings.getSimpleName

abstract class Transformer : Transforms(), PLConstants {


    var _name: String? = null

    fun getName(): String? {
        if (_name == null) {
            _name = getSimpleName(this)
        }
        return _name
    }

//    fun howToHandle(expIn: Exp): Any? {
//        return null
//    }


    abstract fun transform(inExp: Exp): Exp

    fun transformUntilStable(expIn: Exp): Exp {
        return transformUntilStable(this, expIn)
    }

    fun transformArgs(args: List<Exp>): List<Exp> {
        val exps = transformArgs(this, args) ?: throw IllegalStateException()
        return exps
    }

    fun transformArgs2(args: Iterable<Exp>): Iterable<Exp> {
        val exps = transformArgs2(this, args) ?: throw IllegalStateException()
        return exps
    }

    /**
     * Leave Exp.type the same
     */
    protected fun transformArgsOnly(expIn: Exp): Exp {
        checkNotNull(expIn)
        if (expIn.isConstant || expIn.isLit) {
            return expIn
        } else if (expIn.isNegComplex) {
            val argBefore = expIn.arg
            val argAfter = this.transform(argBefore)
            return if (argAfter === argBefore) {
                expIn
            } else {
                argAfter.flip
            }
        } else if (expIn.isPosComplex) {
            val inArgs = expIn.args
            val outArgs = transformArgs2(inArgs)

            val retVal: Exp
            if (outArgs === inArgs) {
                retVal = expIn
            } else {
                retVal = expIn.asPosComplex.newComplex(outArgs)
            }

            checkNotNull(retVal)

            return retVal
        } else {
            throw IllegalArgumentException()
        }


    }

    companion object {

        val FLATTEN_ANDS: Transformer = FlattenAnds()
        val FLATTEN_IMPS: Transformer = FlattenImps()
        val FLATTEN_RMPS: Transformer = FlattenRmps()
        val FLATTEN_ORS: Transformer = FlattenOrs()
        val OR_TO_AND: Transformer = OrToAnd()
        val AND_TO_BINARY: Transformer = AndsToBinary()
        val FLATTEN: Transformer = UntilStableTransformer(FLATTEN_ANDS, FLATTEN_ORS)


        internal var outMatchesInCount = 0

        fun transformUntilStable(transformer: Transformer, expIn: Exp): Exp {

            val out = transformer.transform(expIn)

            return if (out === expIn) {
                out
            } else {
                transformUntilStable(transformer, out)
            }
        }

        fun transformList(transformer: Transformer, list: List<Exp>): List<Exp> {
            return ListTransform.transform(list, transformer)
        }

        fun transformArgs(transformer: Transformer, argsIn: List<Exp>): List<Exp>? {
            val listOut = ListTransform.transform(argsIn, transformer)
            return if (argsIn === listOut) {
                argsIn
            } else {
                listOut
            }
        }

        fun transformArgs2(transformer: Transformer, argsIn: Iterable<Exp>): Iterable<Exp>? {
            val listOut = ListTransform.transform2(argsIn, transformer)
            return if (argsIn === listOut) {
                argsIn
            } else {
                listOut
            }
        }

        //    public ExpSet applyDelta(Exp[] delta) {
        //        if (delta == null) {
        //            return this;
        //        }
        //
        //        Exp[] ret = new Exp[size()];
        //        for (int i = 0; i < size(); i++) {
        //            if (delta[i] != null) {
        //                ret[i] = delta[i];
        //            } else {
        //                ret[i] = get(i);
        //            }
        //        }
        //
        //
        //        return new NSet(ret);
        //    }


        val IDENTITY: Transformer = IdentityTransformer()


        val PUSH_NOTS_IN: Transformer = UntilStableTransformer(CompoundTransformer(
                PushNotsIn(),
                FLATTEN,
                OneArgOr(), OneArgAnd()
        ))

        val PUSH_ANDS_OUT: Transformer = PushAndsOut()


        /**
         * BNF:mBase Normal Form
         * Allow:
         * And(N)
         * Not
         * Or(N)
         * Var
         */
        val BNF: Transformer = UntilStableTransformer("BNF", CompoundTransformer(
                FlattenImps(),
                FlattenRmps(),
                XorsToCnf(),
                ConflictsToCnf(),
                IffToCnf(),
                ImpsToCnf(),
                RmpsToCnf(),
                FlattenOrs(),
                FlattenAnds(),
                OneArgOr(),
                OneArgAnd()
        ))

        val BNF_KEEP_XORS: Transformer = UntilStableTransformer("BNF", CompoundTransformer(
                FlattenImps(),
                FlattenRmps(),
                ConflictsToCnf(),
                IffToCnf(),
                ImpsToCnf(),
                RmpsToCnf(),
                FlattenOrs(),
                FlattenAnds(),
                OneArgOr(),
                OneArgAnd()
        ))


        val BNF_TO_NNF: Transformer = UntilStableTransformer("BNF_TO_NNF", PUSH_NOTS_IN, FLATTEN)

        /**
         * Only allow:
         * Same as BNF,but only _complexVars can be negated
         */
        val NNF: Transformer = UntilStableTransformer("NNF", BNF, BNF_TO_NNF)

        val BNF_TO_AIG: Transformer = UntilStableTransformer("AIG",
                OR_TO_AND,
                AND_TO_BINARY
        )


        val NNF_TO_CNF: Transformer = UntilStableTransformer("NNF_TO_CNF",
                PushAndsOut(),
                OneArgOr(),
                OneArgAnd(),
                FLATTEN
        )

        val CNF: Transformer = UntilStableTransformer("CNF", NNF_TO_CNF, NNF)

        val AIG: Transformer = CompoundTransformer("AIG", BNF, BNF_TO_AIG)
    }


}
